### ISSUE TITLE
Make E2E tests less flaky

### ISSUE THREAD

**Issue by klntsky** [Plutonomicon/cardano-transaction-lib#986]
Remove all timeouts and replace them with functions that continuously query for selector availability.

**Comment by klntsky** [comment]
I was thinking how to make E2E tests less flaky and more easy to use.
There are a few problems with them:

-    it's impossible to know how much time Contract operations will take. But it's required, because we need to have some timeout.
-    It's impossible to know how many of each type of operations would be needed. In our case, we just hardcode wallet operations (1 "allow access" and 1 "sign" for every test except AlwaysSucceeds, which requires two signatures). But our users may not know in advance, because the scenarios they test may have different execution paths.

tl;dr tests are time-based, but app logic is scenario-based.I think we should address this problem by introducing "hooks" that fire before/after common operations, like contract initialization/finalization, signing, submitting, etc.
These hooks will be simple Effect actions, possibly with arguments. They will be configurable through ConfigParams. With hooks, we will be able to allow E2E tests to react on contract state changes, rather than just wait for them with a timeout.
The advantage would be: simpler test logic (basically no need to write modules like Test.E2E.Examples.AlwaysSucceeds), no need to think about delays at all.
